### PR TITLE
Bug 1339547 - Basic set of selenium tests for treeherder

### PR DESCRIPTION
**Pull request by wlach**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/2170)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -148,6 +148,44 @@ matrix:
...
+        # Create a clean virtualenv rather than using the one given to us,
+        # to work around: https://github.com/travis-ci/travis-ci/issues/4873
+        - if [[ ! -f ~/venv/bin/activate ]]; then virtualenv -p python ~/venv; fi
+        - source ~/venv/bin/activate
+        - pip install --disable-pip-version-check --upgrade pip==8.1.1
+        - ./bin/vendor-libmysqlclient.sh ~/venv
+        - wget https://github.com/mozilla/geckodriver/releases/download/v0.14.0/geckodriver-v0.14.0-linux64.tar.gz
+        - mkdir geckodriver
+        - tar -xzf geckodriver-v0.14.0-linux64.tar.gz -C geckodriver
+        - export PATH=$PATH:$PWD/geckodriver
```
Comment (edmorley) [inline .travis.yml]: How about:
```shell
- mkdir $HOME/bin
- curl -sSLf https://github.com/mozilla/geckodriver/releases/download/v0.14.0/geckodriver-v0.14.0-linux64.tar.gz | tar -zxfC $HOME/bin
```
(`$HOME/bin` and `$HOME/.local/bin` are both on `PATH` already; streaming curl into tar means the extraction can happen in parallel)

**Comment by davehunt** [comment]
Review status: 0 of 6 files reviewed at latest revision, 9 unresolved discussions.

---

*[tests/conftest.py, line 46 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3RvWNr-tMsRiW4AU6:-Kd3RvWO3V8Im4X_dper:bo0t1nb) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/conftest.py#L46)):*
> ```Python
> 
>     if 'selenium' in item.keywords and not item.config.getoption("--runselenium"):
>         pytest.skip("need --runselenium option to run selenium tests")
> ```

What's the reason for this command line argument? The --driver argument is already required for running the selenium tests, and it would seem that putting these behind a flag would make them more likely to be missed.

---

*[tests/selenium/test_basics.py, line 10 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3SdBmPRHybO6hWW0b:-Kd3SdBncZoziA5VOjMz:btwynmm) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L10)):*
> ```Python
> @pytest.fixture
> def initial_data(transactional_db):
>     call_command('load_initial_data')
> ```

If all tests are going to require this fixture, then you could consider making it autouse, though we'd probably want to add some logic so we don't try to do this if we run the suite against a deployed instance such as staging. See http://docs.pytest.org/en/latest/fixture.html#autouse-fixtures-xunit-setup-on-steroids

---

*[tests/selenium/test_basics.py, line 13 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3SO5E-_rveKMz0u6K:-Kd3SO5F22XtiMDI1qer:bu1brrv) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L13)):*
> ```Python
> 
> 
> @pytest.mark.selenium
> ```

If you want to mark entire modules, you can: http://docs.pytest.org/en/latest/example/markers.html#marking-whole-classes-or-modules

---

*[tests/selenium/test_basics.py, line 14 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3Stj_4IVASzZJAEPa:-Kd3Stj_4IVASzZJAEPb:bh5wrtl) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L14)):*
> ```Python
> 
> @pytest.mark.selenium
> def test_treeherder_main(initial_data, live_server, selenium):
> ```

Rather than using `live_server` directly, I would suggest implementing a `base_url` fixture override. This could determine if your attempting to run the tests against a deployed instance, and if not it can invoke the `live_server` fixture and return the URL for the test to use.

---

*[tests/selenium/test_basics.py, line 21 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3TE3btjagvZhpTPOa:-Kd3TE3btjagvZhpTPOb:b-t1najw) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L21)):*
> ```Python
>     # selenium = webdriver.Firefox()
>     # driver_.server_url = live_server.url
>     # selenium = webdriver()
> ```

Remove the comments.

---

*[tests/selenium/test_basics.py, line 22 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3THrHrUMqbaVPaXRF:-Kd3THrHrUMqbaVPaXRG:bidqvla) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L22)):*
> ```Python
>     # driver_.server_url = live_server.url
>     # selenium = webdriver()
>     selenium.maximize_window()
> ```

You can override the `selenium` fixture in a `conftest.py` within the selenium directory, just like https://github.com/mozilla/treeherder-tests/blob/8267f4d8d1056bc0f9c9e4edce3c610420e8c2b6/tests/conftest.py#L14-L17

---

*[tests/selenium/test_basics.py, line 28 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3ULJiP3Ge3fV8lpLp:-Kd3ULJiP3Ge3fV8lpLq:b-d63g1z) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L28)):*
> ```Python
>     )
>     repo_button.click()
>     assert selenium.find_element_by_id('repo-dropdown').is_displayed()
> ```

Page objects abstract these locators and provide methods/properties available for the page. Here's an example that opens the repositories menu in Treeherder: https://github.com/mozilla/treeherder-tests/blob/8267f4d8d1056bc0f9c9e4edce3c610420e8c2b6/pages/treeherder.py#L153-L154. This is currently used in a test that switches the repo: https://github.com/mozilla/treeherder-tests/blob/e4c3085792ddfde014c6c3942bf4b955fb43523d/tests/test_switch_repo.py

---

*[tests/selenium/test_basics.py, line 41 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3VE3ah0REtUqa5oBR:-Kd3VE3bPXGJNGTqf2ID:b-unj933) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L41)):*
> ```Python
>     )
>     add_test_button.click()
>     assert selenium.find_elements_by_class_name('modal-header')
> ```

Asserting that this returns 1 or more elements present isn't necessarily going to give much confidence. It's better to check the visibility of elements, or even expand this test to interact with more elements. If any are not present or not displayed as expected then the test will fail. We currently do not have test coverage for this in treeherder-tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/2170)*

**Comment by davehunt** [comment]
That was my first time using Reviewable, and I was expecting a summary comment box for the review. I just wanted to add that this is excellent, and I think it would be fine to land this as it is. We could probably then fairly easily migrate the tests from the treeherder-tests repository and bring along some improvements at the same time. 👍

**Comment by rbillings** [comment]
@davehunt did a thorough review- thanks for that. I don't have any additional notes to add.

**Comment by wlach** [comment]
Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions.

---

*[.travis.yml, line 178 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kcy3QtCY6QFkukyIFzM-r2-178:-Kd7p7TRxzq6I2UG9-bY:b-ltg37a) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/.travis.yml#L178)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary><blockquote>

How about:
```shell
- mkdir $HOME/bin
- curl -sSLf https://github.com/mozilla/geckodriver/releases/download/v0.14.0/geckodriver-v0.14.0-linux64.tar.gz | tar -zxfC $HOME/bin
```
(`$HOME/bin` and `$HOME/.local/bin` are both on `PATH` already; streaming curl into tar means the extraction can happen in parallel)
</blockquote></details>

That generates an error (I think the problem is -C). Fixing that, streaming curl doesn't work, I get an error called `curl: (23) Failed writing body`. I've spent 15 minutes on this, I give up. :) I will extract to $HOME/bin though.

---

*[tests/conftest.py, line 46 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3RvWNr-tMsRiW4AU6:-Kd7n5XdX7Zzo_K2q14n:b-etheuh) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/conftest.py#L46)):*
<details><summary><i>Previously, davehunt (Dave Hunt) wrote…</i></summary><blockquote>

What's the reason for this command line argument? The --driver argument is already required for running the selenium tests, and it would seem that putting these behind a flag would make them more likely to be missed.
</blockquote></details>

The main reason is so that they are skipped when running locally, since that doesn't work currently (we'd need to auto install geckodriver and set up an instance of xvfb). Filed a followup to fix this (and add docs): https://bugzilla.mozilla.org/show_bug.cgi?id=1340305

---

*[tests/selenium/test_basics.py, line 10 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3SdBmPRHybO6hWW0b:-Kd7p1DIg41KM6LADt8N:bxmb4lu) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L10)):*
<details><summary><i>Previously, davehunt (Dave Hunt) wrote…</i></summary><blockquote>

If all tests are going to require this fixture, then you could consider making it autouse, though we'd probably want to add some logic so we don't try to do this if we run the suite against a deployed instance such as staging. See http://docs.pytest.org/en/latest/fixture.html#autouse-fixtures-xunit-setup-on-steroids
</blockquote></details>

Yeah unclear what we want to do here right now, I think I'll defer this for now.

---

*[tests/selenium/test_basics.py, line 13 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3SO5E-_rveKMz0u6K:-Kd7oF_oa4IRsB7I5WaP:b-782kta) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L13)):*
<details><summary><i>Previously, davehunt (Dave Hunt) wrote…</i></summary><blockquote>

If you want to mark entire modules, you can: http://docs.pytest.org/en/latest/example/markers.html#marking-whole-classes-or-modules
</blockquote></details>

Awesome, I'll do that.

---

*[tests/selenium/test_basics.py, line 14 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3Stj_4IVASzZJAEPa:-Kd7oHuX_hCMEh4BTDc2:bulz1zi) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L14)):*
<details><summary><i>Previously, davehunt (Dave Hunt) wrote…</i></summary><blockquote>

Rather than using `live_server` directly, I would suggest implementing a `base_url` fixture override. This could determine if your attempting to run the tests against a deployed instance, and if not it can invoke the `live_server` fixture and return the URL for the test to use.
</blockquote></details>

Ok, I filed a bug to allow us to run these tests against a live version of treeherder (https://bugzilla.mozilla.org/show_bug.cgi?id=1340303). We can do this there.

---

*[tests/selenium/test_basics.py, line 21 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3TE3btjagvZhpTPOa:-Kd7oJvxQ-tJI7-zRJHk:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L21)):*
<details><summary><i>Previously, davehunt (Dave Hunt) wrote…</i></summary><blockquote>

Remove the comments.
</blockquote></details>

Done.

---

*[tests/selenium/test_basics.py, line 22 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3THrHrUMqbaVPaXRF:-Kd7owaMcnJFPiLSnjTj:bpm1l7a) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L22)):*
<details><summary><i>Previously, davehunt (Dave Hunt) wrote…</i></summary><blockquote>

You can override the `selenium` fixture in a `conftest.py` within the selenium directory, just like https://github.com/mozilla/treeherder-tests/blob/8267f4d8d1056bc0f9c9e4edce3c610420e8c2b6/tests/conftest.py#L14-L17
</blockquote></details>

We don't actually need this line, I'll just take it out.

---

*[tests/selenium/test_basics.py, line 28 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3ULJiP3Ge3fV8lpLp:-Kd7onxD3M_c7-WZXxoL:b-iqcafp) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L28)):*
<details><summary><i>Previously, davehunt (Dave Hunt) wrote…</i></summary><blockquote>

Page objects abstract these locators and provide methods/properties available for the page. Here's an example that opens the repositories menu in Treeherder: https://github.com/mozilla/treeherder-tests/blob/8267f4d8d1056bc0f9c9e4edce3c610420e8c2b6/pages/treeherder.py#L153-L154. This is currently used in a test that switches the repo: https://github.com/mozilla/treeherder-tests/blob/e4c3085792ddfde014c6c3942bf4b955fb43523d/tests/test_switch_repo.py
</blockquote></details>

Yeah, this looks like a much better approach to use. I think we can do it when we port the existing treeherder-tests over (https://bugzilla.mozilla.org/show_bug.cgi?id=1286686).

---

*[tests/selenium/test_basics.py, line 41 at r2](https://reviewable.io:443/reviews/mozilla/treeherder/2170#-Kd3VE3ah0REtUqa5oBR:-Kd7oZzYiU7BWhBMYdu5:b-gsjk5z) ([raw file](https://github.com/mozilla/treeherder/blob/5d1f732cbadf0d4b7563eb13a3f34fdab86186ba/tests/selenium/test_basics.py#L41)):*
<details><summary><i>Previously, davehunt (Dave Hunt) wrote…</i></summary><blockquote>

Asserting that this returns 1 or more elements present isn't necessarily going to give much confidence. It's better to check the visibility of elements, or even expand this test to interact with more elements. If any are not present or not displayed as expected then the test will fail. We currently do not have test coverage for this in treeherder-tests.
</blockquote></details>

I will add an id to the element and look for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/2170)*

**Pull request merged**